src/core/eval.ml

```diff
@@ -40,122 +40,133 @@ let eta_equality : bool ref = Console.register_flag "eta_equality" false
-    | (Meta(m1,a1), Meta(m2,a2)) when m1 == m2 ->
+    | (Abst(_ ,b), t | t, Abst(_ ,b)) when !eta_equality ->
+        let x,b = Bindlib.unbind b in
+        eq_modulo ((b, mk_Appl(t, mk_Vari x))::l)
+    | Appl(t1,u1), Appl(t2,u2) -> eq_modulo ((u1,u2)::(t1,t2)::l)
+    | Meta(m1,a1), Meta(m2,a2) when m1 == m2 ->
         eq_modulo (if a1 == a2 then l else List.add_array2 a1 a2 l)
-    | (_          , _          ) -> raise Exit
+    | _ -> raise Exit
   in
-  let res = try eq_modulo [(a,b)]; true with Exit -> false in
-  if !log_enabled && not res then log_conv "failed"; res
+  fun a b ->
+  if !log_enabled then log_conv "%a ≟ %a" pp_term a pp_term b;
+  try eq_modulo [(a,b)]; true
+  with Exit -> if !log_enabled then log_conv "failed"; false
+
+(** [Configuration of the reduction engine. *)
+type config =
+  { context : ctxt (** Context of the reduction used for generating metas. *)
+  ; defmap : term VarMap.t (** Variable definitions. *)
+  ; rewrite : bool (** Use user-defined rewrite rules. *)
+  ; problem : problem (** Generated metavariables. *) }
+
+(*let pp_defmap = D.map VarMap.iter pp_var " ≔ " pp_term "; "*)
+
+let cfg_of_ctx : ctxt -> bool -> config = fun context rewrite ->
+  {context; defmap = Ctxt.to_map context; rewrite; problem = new_problem()}
+
+(** Abstract machine stack. *)
+type stack = term list
+
+(** [appl_to_tref t] transforms {!constructor:Appl} into
+   {!constructor:TRef}. *)
+let appl_to_tref : term -> term = fun t ->
+  match t with
+  | Appl _ -> mk_TRef(ref (Some t))
+  | t -> t
+
+(** [whnf c t] computes a whnf of the term [t] wrt configuration [c]. *)
+let rec whnf : config -> term -> term = fun c t ->
+  (*if !log_enabled then log_eval "whnf %a" pp_term t;*)
+  let s = Stdlib.(!steps) in
+  let u, stk = whnf_stk c t [] in
+  let r = if Stdlib.(!steps) <> s then add_args u stk else unfold t in
+  (*if !log_enabled then log_eval "whnf %a%a ≡ %a"
+    pp_ctxt c.context pp_term t pp_term r;*) r
+
+(** [whnf_stk ~rewrite c t stk] computes a whnf of [add_args t stk] wrt
+   configuration [c]. *)
+and whnf_stk : config -> term -> stack -> term * stack = fun c t stk ->
+  (*if !log_enabled then
+    log_eval "whnf_stk %a %a"
+      pp_ctxt c.context pp_term t (D.list pp_term) stk;*)
+  match unfold t, stk with
+  | Appl(f,u), stk -> whnf_stk c f (appl_to_tref u::stk)
+  | Abst(_,f), u::stk ->
+      Stdlib.incr steps; whnf_stk c (Bindlib.subst f u) stk
+  | LLet(_,t,u), stk ->
+      Stdlib.incr steps; whnf_stk c (Bindlib.subst u (appl_to_tref t)) stk
+  | (Symb s, stk) as r when c.rewrite ->
+      begin match !(s.sym_def) with
+      | Some t when not s.sym_opaq -> Stdlib.incr steps; whnf_stk c t stk
+      | _ ->
+      match tree_walk c !(s.sym_dtree) stk with
+      | None -> r
+      | Some(t,stk) -> Stdlib.incr steps; whnf_stk c t stk
+      end
+  | (Vari x, stk) as r ->
+      begin match VarMap.find_opt x c.defmap with
+      | Some v -> Stdlib.incr steps; whnf_stk c v stk
```

Comment:
@gabrielhdt Note that the normalization of definitions is not shared. We may want to improve that too.